### PR TITLE
API for LoadBalancing Strategy in EnvoyProxy API

### PR DESCRIPTION
**Pull request by arkodg**
* Allows the user to configure a global load balancing strategy for the envoy proxy data plane

```
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: set-lb-strategy
spec:
  traffic:
    loadBalancer:
      strategy: LeastRequest
```
* Sets the default load balancing strategy to `LeastRequest` which will internally set the `WeightedLeastRequest` knob in Envoy

* Relates to https://github.com/envoyproxy/gateway/issues/1256 allowing the user to fall back to `ClusterIP` style loadbalancing

* Relates to https://github.com/envoyproxy/gateway/issues/1105

### CODE REVIEW

**Comment by zirain** [comment]
does gateway API support config this per route/backend?

File: api/config/v1alpha1/envoyproxy_types.go [hunk 1]
```diff
@@ -164,6 +167,49 @@ const (
...
+	LoadBalancerStrategyRoundRobin LoadBalancerStrategy = "RoundRobin"
+
+	// LoadBalancerStrategyRandom defines the "Random" load balancing strategy.
+	// In this mode, the upstream host is chosen randomly.
+	LoadBalancerStrategyRandom LoadBalancerStrategy = "Random"
+
+	// LoadBalancerStrategyClusterIP defines the "ClusterIP" load balancing strategy.
+	// In this mode, EnvoyProxy routes requests to the ClusterIP of the service and
+	// relies on another entity such as kube-proxy to perform load balancing to the upstream hosts.
+	LoadBalancerStrategyClusterIP LoadBalancerStrategy = "ClusterIP"
```
Comment (zirain) [inline api/config/v1alpha1/envoyproxy_types.go:210]: does this mean, eds will just send ClusterIP?

Comment (arkodg) [reply api/config/v1alpha1/envoyproxy_types.go:210]: yes, this is what is happening as of v0.4.0

**Comment by arkodg** [comment]
> does gateway API support config this per route/backend?

no, another way to expose this is to create a hierarchical policy called `TrafficPolicy` that can be applied to Gateway and to HTTPRoute, I think either should work

File: api/config/v1alpha1/envoyproxy_types.go (hunk 1 above)
Comment (zirain) [reply api/config/v1alpha1/envoyproxy_types.go:210]: correct me if I'm wrong, only `ClusterIP` will change the result of `EDS`, other types only change `LbPolicy` on envoy cluster?

Comment (arkodg) [reply api/config/v1alpha1/envoyproxy_types.go:210]: yeah thats right :)

Comment (zirain) [reply api/config/v1alpha1/envoyproxy_types.go:210]: this's a little odd to me, Is there any way to make them more consistent?

Comment (arkodg) [reply api/config/v1alpha1/envoyproxy_types.go:210]: can you share why its odd, this API is meant for end users (platform owners / app owners) to configure LB policy, so they can
* Use LB Policies provided at a L7 request level by Envoy (round-robin, least request, etc) or
* LB Policies provided at a L3/L4 flow level by another intermediate data plane entity such as kube-proxy

**Comment by arkodg** [comment]
closing in favor of https://github.com/envoyproxy/gateway/issues/1492

**Pull request closed without merging**